app/assets/scripts/components/documents/pdf-preview/index.js

```diff
@@ -69,28 +47,76 @@ class AfterRenderHandler extends Handler {
     }
 
     const tocElement = content.querySelector('#table-of-contents');
-    const sectionHeadings = content.querySelectorAll('h2');
-
-    Array.from(sectionHeadings).forEach((h) => {
-      const section = document.createElement('div');
-      section.classList.add('toc-section');
-      tocElement.append(section);
-
-      const sectionTitle = getTitleElement(h.id, h.innerText);
-      section.append(sectionTitle);
-
-      const parent = h.parentNode;
-      const subHeadings = parent.querySelectorAll('h3');
-      if (subHeadings.length > 0) {
-        const subHeadingsContainer = document.createElement('div');
-        subHeadingsContainer.classList.add('sub-headings-container');
-        section.append(subHeadingsContainer);
-        Array.from(subHeadings).forEach((sh) => {
-          const subHeading = getTitleElement(sh.id, sh.innerText);
-          subHeadingsContainer.append(subHeading);
+
+    function generateSubHeadings(
+      currentHeading,
+      currentLevel,
+      sectionContainer,
+      parentContainer,
+      parentHeadingNumber
+    ) {
+      const currentSubHeadings = [];
+      let currentEl = currentHeading;
+
+      while (
+        currentEl.nextElementSibling !== null &&
+        currentEl.nextElementSibling.tagName !== `H${currentLevel}`
+      ) {
+        if (currentEl.nextElementSibling.tagName === `H${currentLevel + 1}`) {
+          currentSubHeadings.push(currentEl.nextElementSibling);
+        }
+
+        currentEl = currentEl.nextElementSibling;
+      }
+
+      if (currentSubHeadings.length > 0) {
+        const subHeadingSections = document.createElement('div');
+        subHeadingSections.classList.add('toc-section');
+        sectionContainer.append(subHeadingSections);
+
+        currentSubHeadings.forEach((subHeading, i) => {
+          const headingNumber = `${parentHeadingNumber}${i + 1}.`;
+          if (subHeading.children[0]) {
+            subHeading.children[0].prepend(`${headingNumber} `);
+          } else {
+            subHeading.prepend(`${headingNumber} `);
+          }
```

Comment:
@batpad This is where the heading number gets added / injected

Reply:
@frozenhelium this all looks really neat to me 👌 . There is quite a lot going on here though, and it's a bit of work to parse through. It would be super helpful to have some comments on this part of the code just describing what it's doing - "Getting all headers, looping through siblings, prepending section number, etc.." .

Apart from that, this looks like like a really neat implementation to me. Let's wait for @wrynearson and @sunu to test before merging.

Reply:
Thanks @frozenhelium, it's amazing that you addressed so many issues with this PR! I'll review this in the next few hours.

Reply:
@batpad I don't have the local backend set up, so once @sunu's [comment](https://github.com/NASA-IMPACT/nasa-apt-frontend/pull/418#pullrequestreview-1336374996) is addressed, we could merge it to staging and I can test there if that works for everyone